content/blog/posts/cmake-essentials-series-part-1.md

```diff
@@ -0,0 +1,75 @@
+---
+title: "CMake Essentials Series - Part 1"
+date: 2022-09-14
+slug: cmake-essentials-series-part-1
+author: Tom Hulton-Harrop
+blog_img: "/images/blog/announcement_thumbnail.jpg"
+full_img: ""
+---
+## Topic
+
+This series aims to give an overview of some of the most useful functionality in CMake and how to apply it.
+
+## Motivation
+
+CMake is the most widely used build system in the C++ community and understanding how best to use it unlocks a whole host of possibilities when it comes to C++ development. Project setup becomes a breeze and integration with open source projects is made much simpler.
+
+## Example
+
+To begin this series we’ll show the minimum amount of code required to get a `‘Hello, World!’` application up and running. First create a new folder and in it create an empty `main.cpp` and `CMakeLists.txt` file.
+
+```bash
+mkdir cmake-essentials-part1 && cd cmake-essentials-part1
+touch main.cpp && touch CMakeLists.txt
+```
+
+The `CMakeLists.txt` file looks like this:
+
+```bash
+cmake_minimum_required(VERSION 3.15) # 1
+project(cmake-essentials-part1 LANGUAGES CXX) # 2
+add_executable(${PROJECT_NAME}) # 3
+target_sources(${PROJECT_NAME} PRIVATE main.cpp) # 4
+target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_17) # 5
+```
+
+First we set the CMake version (`#1`). This line always has to come first in a top-level CMakeLists.txt file (`3.15` is a safe bet but feel free to pick a newer version). Next is the `project` command (`#2`), this should immediately follow `cmake_minimum_required` (`LANGUAGES` is optional but good practice). We then create the target to build (`#3`), in this case an executable (`${PROJECT_NAME}` maps to `cmake-essentials-part1`, the name set in the `project` command). Following that we set the sources to be built (`#4`). As with all* `target_` commands me must specify the scope of the items - here we use `PRIVATE` as these files are not going to be relied on by some downstream dependency. Finally we set the C++ version (`#5`), which again is not strictly required but good practice (note we’re not touching compiler flags here, CMake will handle this for us).
+
+*technically not _all,_ but it’s now best practice to always provide this even if legacy commands do work without it.
+
+`main.cpp` should look familiar:
+
+```c++
+#include <iostream>
+int main(int argc, char argv[]) {
+    std::cout << "Hello, World!\n";
+    return 0;
+}
+```
+
+With that setup, the last thing to do is to run CMake.
+
+```bash
+> cmake -S . -B build # 1
+> cmake --build build # 2
+```
+
+The first line (`#1`) will run the CMake configure step to generate host build files (on Windows this will likely default to Visual Studio and can be specified with `-G`). The following line (`#2`) will invoke whatever build system the configure step generated and leave your application in the `build` folder (with Visual Studio this will be `build/Debug/cmake-essentials-part1.exe`).
```

Comment:
```suggestion
The first line (`#1`) will run the CMake configure step to generate host build files. On Windows this will likely default to Visual Studio and can be specified with `-G`.

The following line (`#2`) will invoke whatever build system the configure step generated and leave your application in the `build` folder. With Visual Studio this will be `build/Debug/cmake-essentials-part1.exe`.
```